<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>gargammel by grenaud</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-dark.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js"></script>
    <script src="javascripts/respond.js"></script>
    <!-- Start of StatCounter Code for Default Guide -->
<script type="text/javascript">
var sc_project=11002891; 
var sc_invisible=1; 
var sc_security="28b7881d"; 
var scJsHost = (("https:" == document.location.protocol) ?
"https://secure." : "http://www.");
document.write("<sc"+"ript type='text/javascript' src='" +
scJsHost+
"statcounter.com/counter/counter.js'></"+"script>");
</script>
<noscript><div class="statcounter"><a title="shopify
analytics ecommerce tracking"
href="http://statcounter.com/shopify/" target="_blank"><img
class="statcounter"
src="//c.statcounter.com/11002891/0/28b7881d/1/"
alt="shopify analytics ecommerce
tracking"></a></div></noscript>
<!-- End of StatCounter Code for Default Guide -->
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!--[if lt IE 8]>
    <link rel="stylesheet" href="stylesheets/ie.css">
    <![endif]-->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

  </head>
  <body>
      <div id="header">
        <nav>
          <li class="fork"><a href="https://github.com/grenaud/gargammel">View On GitHub</a></li>
          <li class="downloads"><a href="https://github.com/grenaud/gargammel/zipball/master">ZIP</a></li>
          <li class="downloads"><a href="https://github.com/grenaud/gargammel/tarball/master">TAR</a></li>
          <li class="title">DOWNLOADS</li>
        </nav>
      </div><!-- end header -->

    <div class="wrapper">

      <section>
        <div id="title">
          <h1>gargammel</h1>
          <p>sequence simulator for ancient DNA</p>
          <hr>
<!--          <span class="credits left">Project maintained by <a href="https://github.com/grenaud">grenaud</a></span> -->
<!--          <span class="credits right">Hosted on GitHub Pages &mdash; Theme by <a href="https://twitter.com/michigangraham">mattgraham</a></span> -->
        </div>

        <h3>
<a id="welcome-to-github-pages" class="anchor" href="#welcome-to-github-pages" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>About</h3>
<p>gargammel is a pipeline to simulate ancient DNA (aDNA) from a set of known references. gargammel seeks to emulate the <i>in vivo</i> process that leads to the sequencing of aDNA fragments:
<ol>
  <li> First, fragments are collected from a set of reference sequences. These reference sequences are designed to represent the endogenous DNA, the contamination both from the same species (e.g. present-day humans), and/or microbes.
  <li> The patterns of mis-incorporations of ancient DNA fragments can be also accounted for.
  <li> Second, aDNA damage is added to those fragments. Sequencing adapters are appended to form reads of a specific length.
  <li> Finally, sequencing errors with corresponding quality scores are applied to the reads.
</ol> 

The resulting data is a set of Illumina reads that can be used to test certain hypotheses about aDNA. Here is a potential set of questions that could be answered using our pipeline:

<ul>	
<li> Impact of present-day human contamination on various statistics used in population genetics
<li> Influence of high levels of deamination on mapping
<li> Impact of a high number of microbial sequences on alignment to a specific reference
<li> Ability to infer the metagenomic profile of a sample depending on the aDNA fragment length distribution
<li> Accuracy of contamination estimates
</ul>

To simulate Illumina sequencing errors,  we use the <a href="http://www.niehs.nih.gov/research/resources/software/biostatistics/art/">ART</a> package.</p>

	<p>
	  gargammel has been developed in <a href="http://geogenetics.ku.dk/research_groups/palaeomix_group/">Ludovic Orlando</a>'s and <a href="http://geogenetics.ku.dk/research_groups/willerslev/">Eske Willerslev</a>'s research groups at the <a href="http://geogenetics.ku.dk/">Center for GeoGenetics</a> at the <a href="http://www.ku.dk/">University of Copenhagen</a>. The code was implemented by <a href="https://grenaud.github.io/">Gabriel Renaud</a> in collaboration with Kristian Hanghoej.	 
	</p>

	<h3>
<a id="designer-templates" class="anchor" href="#designer-templates" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>News</h3>

<ul>
<li>February 28, 2021: We can now handle circular references.</li>

	  
<li>November 21, 2016: gargammel is published in Bioinformatics!</li>
</ul>

	
<h3>
<a id="designer-templates" class="anchor" href="#designer-templates" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download</h3>

You can either type:

<pre>
git clone --depth 1 https://github.com/grenaud/gargammel.git
</pre>

Or click on the tar/zip links at the top of the page.

<h3>
<a id="designer-templates" class="anchor" href="#designer-templates" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Submodules</h3>

<p>gargammel is composed of the following subprograms:
<ol>	
<li> <b>fragSim</b>: simulation of DNA fragmentation due to degradation. Ability to simulate the DNA composition at the 5'/3' ends of the fragments
<li> <b>deamSim</b>: program to add <i>in silico</i> damage (or deamination) to the DNA fragments
<li> <b>apdtSim</b>: module to transform the fragments (damaged or not) into raw Illumina reads
</ol>

The 3 programs were written in C++. There is driver script in Perl that automates the process and simulates the <i>in vivo</i> process that generates aDNA fragments and calls <a href="http://www.niehs.nih.gov/research/resources/software/biostatistics/art/">ART</a> to add sequencing errors. 

</p>

<h3>
<a id="designer-templates" class="anchor" href="#designer-templates" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Documentation, requirements and examples of usage</h3>

<p>For installation, usage and other questions please refer to the README.</p>

<h3>
<a id="creating-pages-manually" class="anchor" href="#creating-pages-manually" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Citing</h3>

<p>
  Please cite our paper:<BR>
  Renaud, G., Hanghoej, K., Willerslev, E. &  Orlando, L. (2016). gargammel: a sequence simulator for ancient DNA <i>Bioinformatics</i>, btw670
</p>



<h3>
<a id="creating-pages-manually" class="anchor" href="#creating-pages-manually" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Acknowledgments</h3>

<p>
This work was supported by the Danish Council for Independent Research, Natural Sciences (4002-00152B); the Danish National Research Foundation (DNRF94); the Villum Fonden (miGENEPI), and; Initiative d'Excellence Chaires d’attractivité, Université de Toulouse (OURASI).
</p>




<h3>
<a id="creating-pages-manually" class="anchor" href="#creating-pages-manually" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Support/Feature request/Push requests</h3>

<p>
<p>Please contact <a href="https://grenaud.github.io/">Gabriel Renaud</a> (<a href="https://github.com/grenaud" class="user-mention">@grenaud</a>) for further information:

<center>
<img src="http://grenaud.github.io/myaddr.png"></img>
</center>

</p>




      </section>

    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->

    
  </body>
</html>
